### PR TITLE
[BREAKING][trtllm] feat: Add FP8 refit support for trtllm rollout

### PR DESCRIPTION
**Pull request by shuyixiong**
### What does this PR do?

This PR adds **FP8 refit support** for the trtllm rollout in verl. 

**Summary of changes:**
- **FP8 refit pipeline**: During weight updates, linear weights (e.g., `q_proj`, `k_proj`, `v_proj`, `o_proj`, `gate_proj`, `up_proj`, `down_proj`, etc.) are quantized to FP8 (e4m3, block size 128×128) via `quant_weights_by_name`; embeddings, norms, and lm_head are excluded.
- **Rollout config**: New config `moe_tensor_parallel_size` for TRT-LLM MoE parallelism.
- **Constraint**: FP8 quantization is only supported with `load_format=dummy` (weights come from the trainer refit path, not from disk).

Experiment:  examples/grpo_trainer/run_qwen3-30b_dapo_megatron_fp8_trtllm.sh & 4 GB200 nodes & ISL 2K &OSL 8K  wandb link: https://wandb.ai/nvidia/Qwen3-30B-A3B-DAPO-GB200/runs/b3zq2zmf?nw=nwusershuyix
image
image

## Test

- **CI**: New E2E job `e2e_dapo_trainer_megatron-qwen3` in `.github/workflows/e2e_ppo_grpo_trainer_trtllm.yml` runs DAPO with FP8 trtllm rollout (Qwen3-0.6B).

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [ ] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [ ] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

File: examples/gmpo_trainer/run_dapo_qwen3_moe_30b_vllm_fp8_rollout.sh [hunk 1]
```diff
@@ -0,0 +1,165 @@
+#!/usr/bin/env bash
+set -xeuo pipefail
```
Comment (hchings) [inline examples/gmpo_trainer/run_dapo_qwen3_moe_30b_vllm_fp8_rollout.sh]: Can we put this only in internal branch, and remove it from this MR?

File: examples/gmpo_trainer/run_dapo_qwen3_moe_30b_trtllm_fp8_rollout.sh [hunk 2]
```diff
@@ -0,0 +1,161 @@
+#!/usr/bin/env bash
+set -xeuo pipefail
+
+gen_tp=2
+NNODES=${NNODES:-2}
```
Comment (hchings) [inline examples/gmpo_trainer/run_dapo_qwen3_moe_30b_trtllm_fp8_rollout.sh]: Note: I'd prefer that this be replaced with megatron script once known issues in megatron are solved, give that 30B is MoE.

File: examples/gmpo_trainer/run_dapo_qwen3_moe_30b_vllm_fp8_rollout.sh (hunk 1 above)
Comment (shuyixiong) [reply examples/gmpo_trainer/run_dapo_qwen3_moe_30b_vllm_fp8_rollout.sh]: Done. The internal branch refers to [https://github.com/shuyixiong/verl/tree/fp8_refit](https://github.com/shuyixiong/verl/tree/fp8_refit).

File: examples/grpo_trainer/run_qwen3-30b_dapo_megatron_fp8_trtllm.sh [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+# -----
+# Config
+# -----
+TP=${INFER_TP:-4}
+ACTOR_TP=${ACTOR_TP:-4}
+ACTOR_PP=${ACTOR_PP:-2}
+ACTOR_VPP=${ACTOR_VPP:-2}
+ACTOR_EP=${ACTOR_EP:-2}
+ACTOR_CP=${ACTOR_CP:-1}
+REF_TP=${REF_TP:4}
```
Comment (hchings) [inline examples/grpo_trainer/run_qwen3-30b_dapo_megatron_fp8_trtllm.sh]: -4

Comment (shuyixiong) [reply examples/grpo_trainer/run_qwen3-30b_dapo_megatron_fp8_trtllm.sh]: Thanks for the correction. Changed in this [commit](https://github.com/verl-project/verl/pull/5374/changes/7fb5a5e3377685519de38d205f6d85569f07f6d2).

File: verl/utils/trtllm/trtllm_fp8_utils.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
+# Copyright 2025 Bytedance Ltd. and/or its affiliates
+# Copyright (c) 2025, NVIDIA CORPORATION. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
```
Comment (Superjomn) [inline verl/utils/trtllm/trtllm_fp8_utils.py:5]: Is this file identical to [sglang_fp8_utils.py](https://github.com/verl-project/verl/blob/main/verl/utils/sglang/sglang_fp8_utils.py)? We need some way to share the common code, and only leave the dedicated changes here.

Comment (shuyixiong) [reply verl/utils/trtllm/trtllm_fp8_utils.py:5]: Using a common `FP8QuantizerHelper` to abstract the implementation. The refactor is included in this [commit](https://github.com/verl-project/verl/pull/5374/changes/ed7da95d350b165f81599a34581744cecbbdde0c).

File: verl/workers/config/rollout.py [hunk 5]
```diff
@@ -167,6 +167,8 @@ class RolloutConfig(BaseConfig):
     expert_parallel_size: int = 1
     tensor_model_parallel_size: int = 2
     pipeline_model_parallel_size: int = 1
+    moe_tp_size: Optional[int] = None
```
Comment (hchings) [inline verl/workers/config/rollout.py]: Can we consolidate the moe ep flag names (`moe_tp_size` and `expert_parallel_size`) ? 
`moe_ep_size` is the convention in TRTLLM, and `expert_parallel_size` is what verl has. 

Same for trtllm_async_server.py L141.

File: .github/workflows/e2e_ppo_grpo_trainer_trtllm.yml [hunk 6]
```diff
@@ -203,9 +208,55 @@ jobs:
         run: |
           rm -rf checkpoints
 
+  e2e_dapo_trainer_megatron-qwen3:
```
Comment (hchings) [inline .github/workflows/e2e_ppo_grpo_trainer_trtllm.yml]: nit: add fp8 in test name to make the test purpose more explicit.

File: examples/grpo_trainer/run_qwen3-30b_dapo_megatron_fp8_trtllm.sh [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+REF_TP=${REF_TP:-4}
+REF_PP=${REF_PP:-2}
+REF_VPP=${REF_VPP:-2}
+REF_EP=${REF_EP:-2}
+REF_CP=${REF_CP:-1}
+GEN_MOE_TP=${GEN_MOE_TP:-2}
+GEN_MOE_EP=${GEN_MOE_EP:-2}
+PROJECT_NAME=${PROJECT_NAME:-"Qwen3-30B-A3B-DAPO-GB200"}
+NNODES=${NNODES:-4}
+GPUS_PER_NODE=${GPUS_PER_NODE:-4}
```
Comment (hchings) [inline examples/grpo_trainer/run_qwen3-30b_dapo_megatron_fp8_trtllm.sh:28]: nit: Should we add a comment header in the script that the config is for GB200, as all our other scripts are for H100s?  Users might assume 8 GPUs per node.

File: verl/workers/config/rollout.py (hunk 5 above)
Comment (Superjomn) [inline verl/workers/config/rollout.py]: Can you add documentation/comments for both new knobs? They may have different semantics depending on the rollout backend, so it’d be helpful to clarify that in the code/docstring.

Also, on naming: other configs like tensor_model_parallel_size use fully spelled-out names, while moe_tp_size feels abbreviated and inconsistent. Would you consider renaming it to match the existing style (e.g., moe_tensor_model_parallel_size or similar)?

Comment (shuyixiong) [reply verl/workers/config/rollout.py]: Can we use `expert_parallel_size` for `moe_expert_parallel_size` and pass `moe_tensor_parallel_size` through engine args because this config is specific to trtllm. (Update: keep `expert_parallel_size` as `moe_ep_size` and rename `moe_tp_size` to `moe_tensor_parallel_size` in this [commit](https://github.com/verl-project/verl/pull/5374/changes/c5f54bc0fc833fdece1103b0d2202337b4ccee7c).")

Comment (shuyixiong) [reply verl/workers/config/rollout.py]: Keep `expert_parallel_size` as `moe_ep_size` and rename `moe_tp_size` to `moe_tensor_parallel_size` in this [commit](https://github.com/verl-project/verl/pull/5374/changes/c5f54bc0fc833fdece1103b0d2202337b4ccee7c).

File: .github/workflows/e2e_ppo_grpo_trainer_trtllm.yml (hunk 6 above)
Comment (shuyixiong) [reply .github/workflows/e2e_ppo_grpo_trainer_trtllm.yml]: Changed in this [commit](https://github.com/verl-project/verl/pull/5374/changes/64e842e4e6152bab0fae5dff42500c284d3b6071).

File: examples/grpo_trainer/run_qwen3-30b_dapo_megatron_fp8_trtllm.sh (hunk 7 above)
Comment (shuyixiong) [reply examples/grpo_trainer/run_qwen3-30b_dapo_megatron_fp8_trtllm.sh:28]: Added in this [commit](https://github.com/verl-project/verl/pull/5374/changes/64e842e4e6152bab0fae5dff42500c284d3b6071).

File: verl/workers/config/rollout.py [hunk 8]
```diff
@@ -167,6 +167,7 @@ class RolloutConfig(BaseConfig):
     expert_parallel_size: int = 1
     tensor_model_parallel_size: int = 2
     pipeline_model_parallel_size: int = 1
+    moe_tensor_parallel_size: int = 1
```
Comment (wuxibin89) [inline verl/workers/config/rollout.py:170]: Should add a field to `verl/trainer/config/rollout/rollout.yaml`

File: verl/utils/fp8_utils.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+import torch
+
+from verl.utils.kernel.fp8_kernel import scaled_fp8_blockwise
+
+logger = logging.getLogger(__file__)
+logger.setLevel(os.getenv("VERL_LOGGING_LEVEL", "INFO"))
+
+
+class FP8QuantizerHelper:
```
Comment (wuxibin89) [inline verl/utils/fp8_utils.py:27]: Is this also common for vllm? @sophiayyya

Comment (sophiayyya) [reply verl/utils/fp8_utils.py:27]: > Is this also common for vllm? @sophiayyya

Yes. I am refactoring vllm_fp8_utils.py to reuse some code and also fix the code break in vllm 0.15.
